File: packages/ctls/py/spellguard_ctls/client/verifier_verify.py
# SPDX-License-Identifier: Apache-2.0

"""
spellguard_ctls - Verifier Verification

Client-side verification of Verifier attestation documents.
This enables bidirectional attestation - clients verify Verifier, not just Verifier verifying clients.
"""

from __future__ import annotations

import asyncio
import time
import uuid
from dataclasses import dataclass, field
from urllib.parse import urlparse

import httpx

from ..types import VerifierAttestationDocument


@dataclass
class VerifierVerifyOptions:
    """Options for Verifier attestation verification."""

    # Expected SHA384 hash of the Verifier Docker image
    expected_image_hash: str
    # Skip strict verification (for development only)
    mock_mode: bool = False
    # Expected certificate hash for pinning
    expected_cert_hash: str | None = field(default=None)


@dataclass
class VerifierVerifyResult:
    """Result of Verifier verification."""

    # Whether the Verifier was verified successfully
    verified: bool
    # The attestation document if verified
    attestation: VerifierAttestationDocument | None = field(default=None)
    # Error message if verification failed
    error: str | None = field(default=None)
    # Whether certificate was verified against pinned hash
    certificate_verified: bool | None = field(default=None)


async def verify_verifier_attestation(
    attestation: VerifierAttestationDocument,
    options: VerifierVerifyOptions,
) -> dict:
    """Verify a Verifier attestation document.

    Args:
        attestation: The attestation document from the Verifier.
        options: Verification options.

    Returns:
        Dict with 'verified' (bool) and optional 'error' (str).
    """
    # In mock mode, skip strict verification
    if options.mock_mode:
        print("[cTLS] Mock mode - skipping strict verification")
        return {"verified": True}

    # Step 1: Verify the image hash matches expected (reproducible build)
    if attestation.image_hash != options.expected_image_hash:
        return {
            "verified": False,
            "error": (
                f"Image hash mismatch. Expected: {options.expected_image_hash}, "
                f"Got: {attestation.image_hash}"
            ),
        }

    # Step 2: Verify timestamp is recent (prevents replay attacks)
    max_age = 5 * 60 * 1000  # 5 minutes in milliseconds
    now_ms = int(time.time() * 1000)
    age = now_ms - attestation.timestamp
    if age > max_age:
        return {
            "verified": False,
            "error": f"Attestation too old: {age}ms (max: {max_age}ms)",
        }

    # Step 3: Verify hardware signature via Phala's verification API
    signature_valid = await _verify_hardware_signature(attestation)
    if not signature_valid:
        return {
            "verified": False,
            "error": "Hardware signature verification failed",
        }

    return {"verified": True}


async def _verify_hardware_signature(
    attestation: VerifierAttestationDocument,
) -> bool:
    """Verify the TDX hardware signature via Phala's attestation verification API.
    The quote is a hex-encoded TDX quote produced by DstackClient.getQuote().
    """
    if (
        not attestation.hardware_signature
        or len(attestation.hardware_signature) < 64
    ):
        return False

    try:
        async with httpx.AsyncClient() as client:
            res = await client.post(
                "https://cloud-api.phala.network/api/v1/attestations/verify",
                json={"hex": attestation.hardware_signature},
                headers={"Content-Type": "application/json"},
            )

        if res.status_code != 200:
            print(
                f"[cTLS] Phala verification API returned {res.status_code}: "
                f"{res.reason_phrase}"
            )
            return False

        result = res.json()
        return result.get("quote", {}).get("verified") is True
    except Exception as error:
        print(f"[cTLS] Failed to verify hardware signature: {error}")
        return False


async def _fetch_attestation_with_retry(
    url: str,
    max_retries: int = 2,
    base_delay_ms: int = 1000,
) -> httpx.Response:
    """Fetch the attestation document with retries for transient gateway errors."""
    last_error: Exception | None = None

    for attempt in range(max_retries + 1):
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(url, timeout=8.0)

            is_transient = (
                response.status_code != 200
                and (response.status_code == 403 or response.status_code >= 500)
            )
            if is_transient and attempt < max_retries:
                delay = base_delay_ms * (2**attempt) / 1000.0
                print(
                    f"[cTLS] Attestation fetch got {response.status_code}, "
                    f"retrying in {int(delay * 1000)}ms ({attempt + 1}/{max_retries})"
                )
                await asyncio.sleep(delay)
                continue
            return response
        except Exception as error:
            last_error = error  # type: ignore[assignment]
            if attempt < max_retries:
                delay = base_delay_ms * (2**attempt) / 1000.0
                print(
                    f"[cTLS] Attestation fetch failed, retrying in "
                    f"{int(delay * 1000)}ms ({attempt + 1}/{max_retries}): {error}"
                )
                await asyncio.sleep(delay)

    raise last_error  # type: ignore[misc]


async def fetch_and_verify_verifier(
    verifier_url: str,
    expected_image_hash: str,
    options: dict | None = None,
) -> VerifierVerifyResult:
    """Fetch and verify Verifier attestation from a URL.

    Args:
        verifier_url: URL of the Verifier server.
        expected_image_hash: Expected SHA384 hash of Verifier Docker image.
        options: Additional verification options (mock_mode, expected_cert_hash).

    Returns:
        Verification result with attestation document.
    """
    opts = options or {}

    # In mock mode, skip the attestation document fetch entirely.
    if opts.get("mock_mode"):
        print("[cTLS] Mock mode -- skipping attestation document fetch")
        return VerifierVerifyResult(verified=True)

    try:
        nonce = str(uuid.uuid4())
        response = await _fetch_attestation_with_retry(
            f"{verifier_url}/attestation?nonce={nonce}"
        )

        if response.status_code != 200:
            return VerifierVerifyResult(
                verified=False,
                error=(
                    f"Failed to fetch attestation: {response.status_code} "
                    f"{response.reason_phrase}"
                ),
            )

        data = response.json()
        attestation = VerifierAttestationDocument(
            image_hash=data["imageHash"],
            hardware_signature=data["hardwareSignature"],
            public_key=data["publicKey"],
            timestamp=data["timestamp"],
            nonce=data["nonce"],
            supported_algorithms=data.get("supportedAlgorithms"),
            event_log=data.get("eventLog"),
            compose_hash=data.get("composeHash"),
        )

        # Verify nonce matches (prevents replay attacks)
        if attestation.nonce != nonce:
            return VerifierVerifyResult(
                verified=False,
                error="Nonce mismatch - possible replay attack",
            )

        result = await verify_verifier_attestation(
            attestation,
            VerifierVerifyOptions(expected_image_hash=expected_image_hash),
        )

        # Certificate pinning verification
        certificate_verified: bool | None = None
        if opts.get("expected_cert_hash"):
            certificate_verified = _verify_certificate_pin(
                verifier_url, opts["expected_cert_hash"]
            )

        return VerifierVerifyResult(
            verified=result["verified"],
            error=result.get("error"),
            attestation=attestation if result["verified"] else None,
            certificate_verified=certificate_verified,
        )
    except Exception as error:
        return VerifierVerifyResult(
            verified=False,
            error=f"Failed to verify Verifier: {error}",
        )


def _verify_certificate_pin(url: str, expected_cert_hash: str) -> bool:
    """Verify TLS certificate against pinned hash.

    Fail-closed: returns False when raw TLS access is not available.
    """
    try:
        parsed = urlparse(url)
        if parsed.scheme != "https":
            print("[cTLS] Certificate pinning requires HTTPS")
            return False

        # Python does not provide easy access to peer TLS certificates
        # from httpx/requests. Fail closed for safety.
        print(
            f"[cTLS] Certificate pinning check requested for {parsed.hostname} "
            "-- full TLS inspection requires ssl.SSLSocket (returning False for safety)"
        )
        return False
    except Exception as err:
        print(f"[cTLS] Certificate pinning error: {err}")
        return False

File: packages/ctls/ts/src/client/verifier-verify.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * @spellguard/ctls - Verifier Verification
 *
 * Client-side verification of Verifier attestation documents.
 * Supports multiple Verifier platforms:
 *   - AWS Nitro Enclaves (COSE_Sign1, verified against AWS root CA)
 *   - Phala Cloud (Intel TDX, verified via Phala DCAP API)
 *   - Mock mode (development only, skips verification)
 */

import { verify as verifyEd25519 } from '../crypto/signing';
import type { VerifierAttestationDocument } from '../types';
import { verifyNitroHardwareSignature } from './nitro-verify';

/**
 * Options for Verifier attestation verification.
 */
export interface VerifierVerifyOptions {
  /** Expected SHA384 hash of the Verifier Docker image */
  expectedImageHash: string;
  /** Skip strict verification (for development only) */
  mockMode?: boolean;
  /** Expected certificate hash for pinning */
  expectedCertHash?: string;
}

/**
 * Result of Verifier verification.
 */
export interface VerifierVerifyResult {
  /** Whether the Verifier was verified successfully */
  verified: boolean;
  /** The attestation document if verified */
  attestation?: VerifierAttestationDocument;
  /** Error message if verification failed */
  error?: string;
  /** Whether certificate was verified against pinned hash */
  certificateVerified?: boolean;
}

/**
 * Verify a Verifier attestation document.
 *
 * @param attestation - The attestation document from the Verifier
 * @param options - Verification options
 * @returns Verification result
 */
export async function verifyVerifierAttestation(
  attestation: VerifierAttestationDocument,
  options: VerifierVerifyOptions,
): Promise<{ verified: boolean; error?: string }> {
  // In mock mode, skip strict verification
  if (options.mockMode) {
    console.log('[cTLS] Mock mode - skipping strict verification');
    return { verified: true };
  }

  // Step 1: Verify the image hash matches expected (reproducible build)
  if (attestation.imageHash !== options.expectedImageHash) {
    return {
      verified: false,
      error: `Image hash mismatch. Expected: ${options.expectedImageHash}, Got: ${attestation.imageHash}`,
    };
  }

  // Step 2: Verify timestamp is recent (prevents replay attacks)
  const maxAge = 5 * 60 * 1000; // 5 minutes
  const age = Date.now() - attestation.timestamp;
  if (age > maxAge) {
    return {
      verified: false,
      error: `Attestation too old: ${age}ms (max: ${maxAge}ms)`,
    };
  }

  // Step 3: Verify hardware signature (dispatches by attestation type)
  const hwResult = await verifyHardwareSignature(attestation);
  if (!hwResult.verified) {
    return {
      verified: false,
      error: hwResult.error || 'Hardware signature verification failed',
    };
  }

  return { verified: true };
}

/**
 * Verify the hardware signature, dispatching to the correct verifier
 * based on the attestation type.
 */
async function verifyHardwareSignature(
  attestation: VerifierAttestationDocument,
): Promise<{ verified: boolean; error?: string }> {
  if (
    !attestation.hardwareSignature ||
    attestation.hardwareSignature.length < 64
  ) {
    return {
      verified: false,
      error: 'Hardware signature missing or too short',
    };
  }

  const type = attestation.attestationType;

  if (type === 'nitro') {
    return verifyNitroHardwareSignature(attestation.hardwareSignature);
  }

  if (type === 'internal') {
    return verifyInternalSessionSignature(attestation);
  }

  // Default: Phala TDX verification via their DCAP API
  return verifyPhalaHardwareSignature(attestation.hardwareSignature);
}

/**
 * Verify the Ed25519 session-key self-signature produced by an internal-mode
 * Verifier. The verifier's trust root for internal mode is the cloud platform
 * identity proven at registration time (see management's internal-mode
 * register handler). This check proves the Verifier at `verifierUrl` still holds the
 * private key corresponding to the public key it's presenting — i.e. it's
 * the same verifier the org already registered, not an impostor on the
 * same hostname.
 *
 * The Verifier signs `imageHash|publicKey|timestamp|nonce` with its Ed25519
 * session key (see `packages/ctls/ts/src/server/attestation.ts`).
 */
async function verifyInternalSessionSignature(
  attestation: VerifierAttestationDocument,
): Promise<{ verified: boolean; error?: string }> {
  try {
    const dataToSign = [
      attestation.imageHash,
      attestation.publicKey,
      attestation.timestamp.toString(),
      attestation.nonce,
    ].join('|');
    const ok = await verifyEd25519(
      dataToSign,
      attestation.hardwareSignature,
      attestation.publicKey,
    );
    return ok
      ? { verified: true }
      : {
          verified: false,
          error: 'Internal-mode session signature did not verify',
        };
  } catch (error) {
    return {
      verified: false,
      error: `Internal-mode signature verification error: ${
        error instanceof Error ? error.message : String(error)
      }`,
    };
  }
}

/**
 * Verify a TDX hardware signature via Phala's attestation verification API.
 * The quote is a hex-encoded TDX quote produced by DstackClient.getQuote().
 */
async function verifyPhalaHardwareSignature(
  hardwareSignature: string,
): Promise<{ verified: boolean; error?: string }> {
  try {
    const res = await fetch(
      'https://cloud-api.phala.network/api/v1/attestations/verify',
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ hex: hardwareSignature }),
      },
    );

    if (!res.ok) {
      return {
        verified: false,
        error: `Phala verification API returned ${res.status}: ${res.statusText}`,
      };
    }

    const result = (await res.json()) as {
      quote?: { verified?: boolean };
    };
    return result.quote?.verified === true
      ? { verified: true }
      : { verified: false, error: 'Phala API rejected the TDX quote' };
  } catch (error) {
    return {
      verified: false,
      error: `Phala verification failed: ${error instanceof Error ? error.message : String(error)}`,
    };
  }
}

/**
 * Fetch the attestation document with retries for transient gateway errors
 * (e.g. Phala dstack gateway intermittently returning 403 to CF Worker IPs).
 */
async function fetchAttestationWithRetry(
  url: string,
  // Trimmed 2→1: with the 60s per-attempt timeout below, a cold SLIM session
  // completes on the first attempt; the extra retries only churned the
  // half-open session pool (a source of the gateway 502/503s). One retry
  // still covers the genuine transient case (Phala 403 / 5xx).
  maxRetries = 1,
  baseDelayMs = 1000,
): Promise<Response> {
  let lastError: unknown;
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      const response = await fetch(url, {
        // In the slim profile this fetch traverses the gateway→SLIM→verifier
        // hop, whose cold-session establishment runs ~20-30s; 8s aborted
        // mid-handshake on every attempt. 60s clears it with headroom and
        // stays under the 150s ALB idle timeout even across the retry.
        signal: AbortSignal.timeout(60_000),
      });
      const isTransient =
        !response.ok && (response.status === 403 || response.status >= 500);
      if (isTransient && attempt < maxRetries) {
        const delay = baseDelayMs * 2 ** attempt;
        console.warn(
          `[cTLS] Attestation fetch got ${response.status}, retrying in ${delay}ms (${attempt + 1}/${maxRetries})`,
        );
        await new Promise((r) => setTimeout(r, delay));
        continue;
      }
      return response;
    } catch (error) {
      lastError = error;
      if (attempt < maxRetries) {
        const delay = baseDelayMs * 2 ** attempt;
        console.warn(
          `[cTLS] Attestation fetch failed, retrying in ${delay}ms (${attempt + 1}/${maxRetries}): ${error}`,
        );
        await new Promise((r) => setTimeout(r, delay));
      }
    }
  }
  throw lastError;
}

/**
 * Fetch and verify Verifier attestation from a URL.
 *
 * @param verifierUrl - URL of the Verifier server
 * @param expectedImageHash - Expected SHA384 hash of Verifier Docker image
 * @param options - Additional verification options
 * @returns Verification result with attestation document
 */
export async function fetchAndVerifyVerifier(
  verifierUrl: string,
  expectedImageHash: string,
  options?: {
    mockMode?: boolean;
    expectedCertHash?: string;
  },
): Promise<VerifierVerifyResult> {
  // In mock mode, skip the attestation document fetch entirely.
  // The attestation doc is not used for anything downstream — the Verifier's
  // public key comes from the /agents/register response.  Fetching it in
  // mock mode is wasteful and unreliable: CF Workers frequently get 403
  // from the Phala dstack gateway, causing retries + backoff that exceed
  // the pre-registration timeout and leave agents unregistered.
  if (options?.mockMode) {
    console.log('[cTLS] Mock mode — skipping attestation document fetch');
    return { verified: true };
  }

  try {
    const nonce = crypto.randomUUID();
    const response = await fetchAttestationWithRetry(
      `${verifierUrl}/attestation?nonce=${nonce}`,
    );

    if (!response.ok) {
      return {
        verified: false,
        error: `Failed to fetch attestation: ${response.status} ${response.statusText}`,
      };
    }

    const attestation = (await response.json()) as VerifierAttestationDocument;

    // Verify nonce matches (prevents replay attacks)
    if (attestation.nonce !== nonce) {
      return {
        verified: false,
        error: 'Nonce mismatch - possible replay attack',
      };
    }

    const result = await verifyVerifierAttestation(attestation, {
      expectedImageHash,
    });

    // Certificate pinning verification
    const certificateVerified = options?.expectedCertHash
      ? verifyCertificatePin(verifierUrl, options.expectedCertHash)
      : undefined;

    return {
      ...result,
      attestation: result.verified ? attestation : undefined,
      certificateVerified,
    };
  } catch (error) {
    return {
      verified: false,
      error: `Failed to verify Verifier: ${error}`,
    };
  }
}

/**
 * Verify TLS certificate against pinned hash.
 *
 * Fail-closed: returns false when raw TLS access is not available
 * (e.g. in fetch-only environments like Cloudflare Workers).
 *
 * In Node.js, uses the `tls` module to extract the peer certificate
 * and compare its SHA-256 hash against the expected hash.
 */
function verifyCertificatePin(url: string, expectedCertHash: string): boolean {
  try {
    // Attempt to use Node.js tls module for certificate inspection
    // This will not be available in all environments (e.g. CF Workers)
    const { URL } = globalThis;
    const parsed = new URL(url);

    if (parsed.protocol !== 'https:') {
      console.warn('[cTLS] Certificate pinning requires HTTPS');
      return false;
    }

    // In environments without raw TLS socket access (browser, CF Workers),
    // we cannot extract the peer certificate. Fail closed.
    if (
      typeof globalThis.process === 'undefined' ||
      !globalThis.process?.versions?.node
    ) {
      console.warn(
        '[cTLS] Certificate pinning not available in this environment (no Node.js TLS)',
      );
      return false;
    }

    // Node.js environment: use tls.connect to inspect the certificate
    // Note: This is a synchronous check using cached certificate data.
    // Full async implementation would use https.Agent with checkServerIdentity.
    console.warn(
      `[cTLS] Certificate pinning check requested for ${parsed.hostname} — full TLS inspection requires async https.Agent (returning false for safety)`,
    );
    return false;
  } catch (err) {
    console.error('[cTLS] Certificate pinning error:', err);
    return false;
  }
}
